

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>1. Testing dataset &mdash; bhive documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="1. BC_edit_matrix.py" href="demo/BC_edit_matrix.html" />
    <link rel="prev" title="1. bhive Release history" href="change_log.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search"  style="background: white" >
          

          
            <a href="index.html" class="icon icon-home"> bhive
          

          
            
            <img src="_static/bhive_logo.png" class="logo" alt="Logo"/>
          
          </a>

          
            
            
              <div class="version">
                1.0.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Overview</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="overview.html">1. Overview</a></li>
</ul>
<p class="caption"><span class="caption-text">Installation</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">1. Installation</a></li>
</ul>
<p class="caption"><span class="caption-text">Change log</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="change_log.html">1. bhive Release history</a></li>
</ul>
<p class="caption"><span class="caption-text">Test Dataset</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">1. Testing dataset</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#download-raw-data">1.1. Download raw data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#convert-bam-to-fastq">1.2. Convert BAM to FASTQ</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-cellranger-count-workflow">1.3. Run CellRanger <em>count</em> workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-cellranger-aggr-workflow">1.4. Run CellRanger <em>aggr</em> workflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="#references">1.5. References</a></li>
</ul>
</li>
</ul>
<p class="caption"><span class="caption-text">Usage Information</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="demo/BC_edit_matrix.html">1. BC_edit_matrix.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/cell_calling.html">2. cell_calling.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/map_stat.html">3. map_stat.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/seq_logo.html">4. seq_logo.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="demo/seq_qual.html">5. seq_qual.py</a></li>
</ul>
<p class="caption"><span class="caption-text">Liscense</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="liscense.html">LICENSE</a></li>
</ul>
<p class="caption"><span class="caption-text">Reference</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="reference.html">Reference</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">bhive</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li><span class="section-number">1. </span>Testing dataset</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/dataset.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="testing-dataset">
<h1><span class="section-number">1. </span>Testing dataset<a class="headerlink" href="#testing-dataset" title="Permalink to this headline">¶</a></h1>
<p>We use the same example dataset as used by <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/tutorials/gex-analysis-nature-publication">10X Genomics</a>.
Raw data (in BAM format) were downloaded from the NCBI Sequence Read Archive (SRA). The study was published in <a class="footnote-reference brackets" href="#id3" id="id1">1</a>.</p>
<div class="section" id="download-raw-data">
<h2><span class="section-number">1.1. </span>Download raw data<a class="headerlink" href="#download-raw-data" title="Permalink to this headline">¶</a></h2>
<table class="docutils align-default">
<colgroup>
<col style="width: 47%" />
<col style="width: 9%" />
<col style="width: 9%" />
<col style="width: 7%" />
<col style="width: 7%" />
<col style="width: 21%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p>File_name</p></td>
<td><p>GSM_accession</p></td>
<td><p>SRR_accession</p></td>
<td><p>File_size</p></td>
<td><p>Treatment</p></td>
<td><p>MD5</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://sra-pub-src-1.s3.amazonaws.com/SRR7611046/C05.bam.1">C05.bam.1</a></p></td>
<td><p>GSM3308718</p></td>
<td><p>SRR7611046</p></td>
<td><p>70 Gb</p></td>
<td><p>normal</p></td>
<td><p>97b87c87b539e69dad7dcb04e8f03132</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://sra-pub-src-1.s3.amazonaws.com/SRR7611048/C07.bam.1">C07.bam.1</a></p></td>
<td><p>GSM3308720</p></td>
<td><p>SRR7611048</p></td>
<td><p>64 Gb</p></td>
<td><p>irradiated</p></td>
<td><p>064669deb6be22e5f82fe58679f7e394</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="convert-bam-to-fastq">
<h2><span class="section-number">1.2. </span>Convert BAM to FASTQ<a class="headerlink" href="#convert-bam-to-fastq" title="Permalink to this headline">¶</a></h2>
<p>Download <code class="docutils literal notranslate"><span class="pre">bamtofastq</span></code> from <a class="reference external" href="https://cf.10xgenomics.com/misc/bamtofastq-1.2.0">here</a>. Convert BAM into FASTQ files.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ bamtofastq C05.bam.1 normal_dat
$ bamtofastq C07.bam.1 irradiated_dat
</pre></div>
</div>
<p>After this step, you will get two subdirectories (<code class="docutils literal notranslate"><span class="pre">./normal_dat</span></code> and <code class="docutils literal notranslate"><span class="pre">./irradiated_dat</span></code>) under your current directory. And within <code class="docutils literal notranslate"><span class="pre">./normal_dat</span></code> and <code class="docutils literal notranslate"><span class="pre">./irradiated_dat</span></code>, there are
subdirectories and fastq files, for example</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd ./normal_dat
$ tree
.
├── indepth_C05_MissingLibrary_1_HL5G3BBXX
│   ├── bamtofastq_S1_L003_I1_001.fastq.gz
│   ├── bamtofastq_S1_L003_I1_002.fastq.gz
│   ├── bamtofastq_S1_L003_R1_001.fastq.gz
│   ├── bamtofastq_S1_L003_R1_002.fastq.gz
│   ├── bamtofastq_S1_L003_R2_001.fastq.gz
│   ├── bamtofastq_S1_L003_R2_002.fastq.gz
│   ├── bamtofastq_S1_L004_I1_001.fastq.gz
│   ├── bamtofastq_S1_L004_I1_002.fastq.gz
│   ├── bamtofastq_S1_L004_I1_003.fastq.gz
│   ├── bamtofastq_S1_L004_I1_004.fastq.gz
│   ├── bamtofastq_S1_L004_I1_005.fastq.gz
│   ├── bamtofastq_S1_L004_I1_006.fastq.gz
│   ├── bamtofastq_S1_L004_R1_001.fastq.gz
│   ├── bamtofastq_S1_L004_R1_002.fastq.gz
│   ├── bamtofastq_S1_L004_R1_003.fastq.gz
│   ├── bamtofastq_S1_L004_R1_004.fastq.gz
│   ├── bamtofastq_S1_L004_R1_005.fastq.gz
│   ├── bamtofastq_S1_L004_R1_006.fastq.gz
│   ├── bamtofastq_S1_L004_R2_001.fastq.gz
│   ├── bamtofastq_S1_L004_R2_002.fastq.gz
│   ├── bamtofastq_S1_L004_R2_003.fastq.gz
│   ├── bamtofastq_S1_L004_R2_004.fastq.gz
│   ├── bamtofastq_S1_L004_R2_005.fastq.gz
│   └── bamtofastq_S1_L004_R2_006.fastq.gz
└── indepth_C05_MissingLibrary_1_HNNWNBBXX
    ├── bamtofastq_S1_L002_I1_001.fastq.gz
    ├── bamtofastq_S1_L002_I1_002.fastq.gz
    ├── bamtofastq_S1_L002_I1_003.fastq.gz
    ├── bamtofastq_S1_L002_I1_004.fastq.gz
    ├── bamtofastq_S1_L002_I1_005.fastq.gz
    ├── bamtofastq_S1_L002_R1_001.fastq.gz
    ├── bamtofastq_S1_L002_R1_002.fastq.gz
    ├── bamtofastq_S1_L002_R1_003.fastq.gz
    ├── bamtofastq_S1_L002_R1_004.fastq.gz
    ├── bamtofastq_S1_L002_R1_005.fastq.gz
    ├── bamtofastq_S1_L002_R2_001.fastq.gz
    ├── bamtofastq_S1_L002_R2_002.fastq.gz
    ├── bamtofastq_S1_L002_R2_003.fastq.gz
    ├── bamtofastq_S1_L002_R2_004.fastq.gz
    ├── bamtofastq_S1_L002_R2_005.fastq.gz
    ├── bamtofastq_S1_L003_I1_001.fastq.gz
    ├── bamtofastq_S1_L003_I1_002.fastq.gz
    ├── bamtofastq_S1_L003_R1_001.fastq.gz
    ├── bamtofastq_S1_L003_R1_002.fastq.gz
    ├── bamtofastq_S1_L003_R2_001.fastq.gz
    └── bamtofastq_S1_L003_R2_002.fastq.gz
</pre></div>
</div>
</div>
<div class="section" id="run-cellranger-count-workflow">
<h2><span class="section-number">1.3. </span>Run CellRanger <em>count</em> workflow<a class="headerlink" href="#run-cellranger-count-workflow" title="Permalink to this headline">¶</a></h2>
<p>Download <code class="docutils literal notranslate"><span class="pre">cellranger</span></code> and <strong>Mouse reference dataset</strong> from <a class="reference external" href="https://support.10xgenomics.com/single-cell-gene-expression/software/downloads/latest">here</a></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cellranger --version
cellranger 4.0.0

# run cellranger for normal sample
$ cd ./normal_dat
$ cellranger count  --id=normal        --transcriptome=/XYZ/CellRanger/refdata-gex-mm10-2020-A  --fastqs=./indepth_C05_MissingLibrary_1_HL5G3BBXX,./indepth_C05_MissingLibrary_1_HNNWNBBXX

# run cellranger for irradiated sample
$ cd ./irradiated_dat
$ cellranger count  --id=irradiated    --transcriptome=/XYZ/CellRanger/refdata-gex-mm10-2020-A  --fastqs=./indepth_C07_MissingLibrary_1_HL5G3BBXX,./indepth_C07_MissingLibrary_1_HNNWNBBXX
</pre></div>
</div>
<p>After each <code class="docutils literal notranslate"><span class="pre">cellranger</span> <span class="pre">count</span></code> workflow is finished successfully. Subdirectories <code class="docutils literal notranslate"><span class="pre">normal</span></code> and <code class="docutils literal notranslate"><span class="pre">irradiated</span></code> will be created, which contain the cellranger outputs. For example,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd normal
$ ls -F
_cmdline     _invocation  _mrosource     _perf               _tags       _vdrkill
_filelist    _jobmode    normal.mri.tgz  SC_RNA_COUNTER_CS/  _timestamp  _versions
_finalstate  _log        outs/           _sitecheck          _uuid
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Replace /XYZ/ with the actual path on your system.</p>
</div>
</div>
<div class="section" id="run-cellranger-aggr-workflow">
<h2><span class="section-number">1.4. </span>Run CellRanger <em>aggr</em> workflow<a class="headerlink" href="#run-cellranger-aggr-workflow" title="Permalink to this headline">¶</a></h2>
<p>First, make the <code class="docutils literal notranslate"><span class="pre">library.csv</span></code> file. This CSV file has two columns which define the <strong>ID</strong> and the location of the <strong>molecule_info.h5</strong> file from each run.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cat  library.csv

library_id,molecule_h5
normal,/ABC/normal_dat/normal/outs/molecule_info.h5
irradiated,/ABC/irradiated_dat/irradiated/outs/molecule_info.h5
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Replace /ABC/ with the actual path on your system.</p>
</div>
<p>Then, run <code class="docutils literal notranslate"><span class="pre">cellranger</span> <span class="pre">aggr</span></code> workflow. The <code class="docutils literal notranslate"><span class="pre">cellranger</span> <span class="pre">aggr</span></code> workflow aggregates outputs from multiple runs of the <code class="docutils literal notranslate"><span class="pre">cellranger</span> <span class="pre">count</span></code> workflow</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cellranger aggr --id=aggr --csv=libraries.csv
</pre></div>
</div>
<p>After each <code class="docutils literal notranslate"><span class="pre">cellranger</span> <span class="pre">aggr</span></code> workflow is finished successfully. A subdirectory <code class="docutils literal notranslate"><span class="pre">aggr</span></code> will be created, which contain the cellranger outputs. For example,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ cd aggr
$ ls -F
aggr.mri.tgz  _finalstate  _log        _perf                 _tags       _vdrkill
_cmdline      _invocation  _mrosource  SC_RNA_AGGREGATOR_CS/  _timestamp  _versions
_filelist     _jobmode    outs/       _sitecheck             _uuid
</pre></div>
</div>
</div>
<div class="section" id="references">
<h2><span class="section-number">1.5. </span>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<dl class="footnote brackets">
<dt class="label" id="id3"><span class="brackets"><a class="fn-backref" href="#id1">1</a></span></dt>
<dd><p>Ayyaz A, Kumar S, Sangiorgi B, Ghoshal B, Gosio J, Ouladan S, Fink M, Barutcu S, Trcka D, Shen J, Chan K, Wrana JL, Gregorieff A. Single-cell transcriptomes of the regenerating intestine reveal a revival stem cell. Nature. 2019 May;569(7754):121-125. doi: 10.1038/s41586-019-1154-y. Epub 2019 Apr 24. PMID: 31019301.</p>
</dd>
</dl>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="demo/BC_edit_matrix.html" class="btn btn-neutral float-right" title="1. BC_edit_matrix.py" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="change_log.html" class="btn btn-neutral float-left" title="1. bhive Release history" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Liguo Wang et al
      <span class="lastupdated">
        Last updated on July 9, 2020.
      </span>

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-XXXXXXX-1', 'auto');
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>